[FX] Fix bare generic type annotations

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #74135

Previously, if a user had annotated a type with `typing.Tuple` (no subscript), the generated code in fx would throw a syntax error like:

```
File "/torch/fx/graph.py", line 346, in type_repr
    args = [type_repr(arg) for arg in o.__args__]
  File "/python3.9/typing.py", line 706, in _getattr_
    raise AttributeError(attr)
AttributeError: _args_
```

or

```
File "<eval_with_key>.0", line 4
    def forward(self, input_ids : typing_Union[torch.Tensor,NoneType]) -> typing_Union[typing_Tuple[],transformers_modeling_outputs_BaseModelOutputWithPoolingAndCrossAttentions]:
```

This fixes that by emitting the same syntax for these type annotation as they were written, i.e. `typing.Tuple` -> `typing.Tuple`

Differential Revision: [D34839339](https://our.internmc.facebook.com/intern/diff/D34839339)

Commit:
[FX] Fix bare generic type annotations

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34839339).

pbelevich commented:
```
Traceback (most recent call last):
  File "/Users/pbelevich/PycharmProjects/PiPPy/test/hf_bert_test.py", line 28, in <module>
    bert_pipe = Pipe.from_tracing(bert, MultiUseParameterConfig.TRANSMIT, tracer=hf_tracer, concrete_args=concrete_args)
  File "/Users/pbelevich/PycharmProjects/PiPPy/pippy/IR.py", line 602, in from_tracing
    traced = torch.fx.GraphModule(mod, graph)
  File "/Users/pbelevich/miniconda3/envs/PiPPy/lib/python3.8/site-packages/torch/fx/graph_module.py", line 312, in __init__
    self.graph = graph
  File "/Users/pbelevich/miniconda3/envs/PiPPy/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1234, in __setattr__
    object.__setattr__(self, name, value)
  File "/Users/pbelevich/miniconda3/envs/PiPPy/lib/python3.8/site-packages/torch/fx/graph_module.py", line 346, in graph
    self.recompile()
  File "/Users/pbelevich/miniconda3/envs/PiPPy/lib/python3.8/site-packages/torch/fx/graph_module.py", line 576, in recompile
    python_code = self._graph.python_code(root_module='self')
  File "/Users/pbelevich/miniconda3/envs/PiPPy/lib/python3.8/site-packages/torch/fx/graph.py", line 1142, in python_code
    return self._python_code(root_module, namespace)
  File "/Users/pbelevich/miniconda3/envs/PiPPy/lib/python3.8/site-packages/torch/fx/graph.py", line 1145, in _python_code
    return self._codegen._gen_python_code(self.nodes, root_module, namespace)
  File "/Users/pbelevich/miniconda3/envs/PiPPy/lib/python3.8/site-packages/torch/fx/graph.py", line 472, in _gen_python_code
    emit_node(node)
  File "/Users/pbelevich/miniconda3/envs/PiPPy/lib/python3.8/site-packages/torch/fx/graph.py", line 407, in emit_node
    maybe_type_annotation = '' if node.type is None else f' : {type_repr(node.type)}'
  File "/Users/pbelevich/miniconda3/envs/PiPPy/lib/python3.8/site-packages/torch/fx/graph.py", line 347, in type_repr
    args = [type_repr(arg) for arg in o.__args__]
  File "/Users/pbelevich/miniconda3/envs/PiPPy/lib/python3.8/site-packages/torch/fx/graph.py", line 347, in <listcomp>
    args = [type_repr(arg) for arg in o.__args__]
  File "/Users/pbelevich/miniconda3/envs/PiPPy/lib/python3.8/site-packages/torch/fx/graph.py", line 340, in type_repr
    origin_type = _origin_type_map.get(o.__origin__, o.__origin__)
AttributeError: type object 'Tensor' has no attribute '__origin__'
```

Commit:
Update on "[FX] Fix bare generic type annotations"


Previously, if a user had annotated a type with `typing.Tuple` (no subscript), the generated code in fx would throw a syntax error like:

```
File "/torch/fx/graph.py", line 346, in type_repr
    args = [type_repr(arg) for arg in o.__args__]
  File "/python3.9/typing.py", line 706, in _getattr_
    raise AttributeError(attr)
AttributeError: _args_
```

or

```
File "<eval_with_key>.0", line 4
    def forward(self, input_ids : typing_Union[torch.Tensor,NoneType]) -> typing_Union[typing_Tuple[],transformers_modeling_outputs_BaseModelOutputWithPoolingAndCrossAttentions]:
```

This fixes that by emitting the same syntax for these type annotation as they were written, i.e. `typing.Tuple` -> `typing.Tuple`

Differential Revision: [D34839339](https://our.internmc.facebook.com/intern/diff/D34839339)

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34839339).

pbelevich commented on torch/fx/graph.py:
```diff
@@ -337,15 +337,19 @@ def type_repr(o : Any):
-                args = [type_repr(arg) for arg in o.__args__]
+                if hasattr(o, '__args__'):
+                    # Assign global names for each of the inner type variables.
+                    args = [type_repr(arg) for arg in o.__args__]
```
```suggestion
                    args = [type_repr(arg) for arg in o.__args__]
                    if len(args) == 0:
                        # Bare type, such as `typing.Tuple` with no subscript
                        return origin_typename
```

pbelevich commented on torch/fx/graph.py:
```diff
@@ -337,15 +337,19 @@ def type_repr(o : Any):
 
-                return f'{origin_typename}[{",".join(args)}]'
+                    return f'{origin_typename}[{",".join(args)}]'
+                else:
```
unreachable?

jamesr66a replied:
No, different Python versions have different behavior. In my tests on Python 3.9, a bare `typing.Tuple` type expression does not have an attribute `__args__`. Pre-3.9 has attribute `__args__`, but it is empty. We will need to keep both this branch and your suggested branch above

jamesr66a replied:
3.9:

```
Python 3.9.5 (default, Jun  4 2021, 12:28:51) 
[GCC 7.5.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing import Tuple
>>> t = Tuple
>>> hasattr(t, '__args__')
False
```

3.8:

```
Python 3.8.12 (default, Oct 12 2021, 13:49:34) 
[GCC 7.5.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing import Tuple
>>> t = Tuple
>>> hasattr(t, '__args__')
True
>>> t.__args__
()
```

Commit:
Update on "[FX] Fix bare generic type annotations"


Previously, if a user had annotated a type with `typing.Tuple` (no subscript), the generated code in fx would throw a syntax error like:

```
File "/torch/fx/graph.py", line 346, in type_repr
    args = [type_repr(arg) for arg in o.__args__]
  File "/python3.9/typing.py", line 706, in _getattr_
    raise AttributeError(attr)
AttributeError: _args_
```

or

```
File "<eval_with_key>.0", line 4
    def forward(self, input_ids : typing_Union[torch.Tensor,NoneType]) -> typing_Union[typing_Tuple[],transformers_modeling_outputs_BaseModelOutputWithPoolingAndCrossAttentions]:
```

This fixes that by emitting the same syntax for these type annotation as they were written, i.e. `typing.Tuple` -> `typing.Tuple`

Differential Revision: [D34839339](https://our.internmc.facebook.com/intern/diff/D34839339)

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34839339).

pbelevich reviewed (approved):
LGTM
